File: pkms-backend/app/models/enums.py
"""
Shared Enums for PKMS Models
"""
import enum

class ModuleType(str, enum.Enum):
    """Enum for different module types in the application."""
    NOTES = "notes"
    DOCUMENTS = "documents"
    TODOS = "todos"
    PROJECTS = "projects"
    DIARY = "diary"
    ARCHIVE_ITEMS = "archive_items"
    ARCHIVE_FOLDERS = "archive_folders"
    GENERAL = "general"

class ProjectStatus(str, enum.Enum):
    """Project status enum"""
    IS_RUNNING = "is_running"
    ON_HOLD = "on_hold"
    COMPLETED = "completed"
    CANCELLED = "cancelled"

class TodoStatus(str, enum.Enum):
    """Todo status enum for better task management"""
    PENDING = "pending"
    IN_PROGRESS = "in_progress"
    BLOCKED = "blocked"
    DONE = "done"
    CANCELLED = "cancelled"

class TodoType(str, enum.Enum):
    """Todo type enum for different kinds of tasks"""
    TASK = "task"
    CHECKLIST = "checklist"
    SUBTASK = "subtask"

class TaskPriority(str, enum.Enum):
    """Priority levels for tasks, todos, and projects"""
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"
    URGENT = "urgent"

class UploadStatus(str, enum.Enum):
    """Document upload status enum"""
    PENDING = "pending"
    PROCESSING = "processing"
    COMPLETED = "completed"
    FAILED = "failed"

class ChunkUploadStatus(str, enum.Enum):
    """Chunked file upload status enum"""
    UPLOADING = "uploading"
    ASSEMBLING = "assembling"
    COMPLETED = "completed"
    FAILED = "failed"
    ERROR = "error"

class TodoStatsKey(str, enum.Enum):
    """Todo statistics keys for dashboard"""
    # Status-based counts (from TodoStatus enum)
    TOTAL = "total"
    PENDING = "pending"
    IN_PROGRESS = "in_progress"
    BLOCKED = "blocked"
    DONE = "done"
    
    # Time-based computed counts (calculated on the fly - active todos only)
    OVERDUE = "overdue"           # due_date < today AND status not DONE/CANCELLED
    DUE_TODAY = "due_today"       # due_date == today AND status not DONE/CANCELLED  
    COMPLETED_TODAY = "completed_today"  # completed_at == today (DONE status only)
    WITHIN_TIME = "within_time"   # due_date >= today AND status not DONE/CANCELLED

class ModuleStatsKey(str, enum.Enum):
    """Module statistics keys for dashboard"""
    TOTAL = "total"
    RECENT = "recent"

File: pkms-frontend/src/types/enums.ts
/**
 * TypeScript Enums matching backend app/models/enums.py exactly
 * Ensures type safety and prevents typos in status/priority comparisons
 * 
 * ARCHITECTURAL_RULES.md Rule #6 (Enum Usage - Type Safety)
 * Rule #10 (Centralized Enum Definitions)
 */

export enum ModuleType {
  NOTES = "notes",
  DOCUMENTS = "documents",
  TODOS = "todos",
  PROJECTS = "projects",
  DIARY = "diary",
  ARCHIVE_ITEMS = "archive_items",
  ARCHIVE_FOLDERS = "archive_folders",
  GENERAL = "general"
}

export enum ProjectStatus {
  IS_RUNNING = "is_running",  // URL param: ?status=is_running
  ON_HOLD = "on_hold",        // URL param: ?status=on_hold
  COMPLETED = "completed",    // URL param: ?status=completed
  CANCELLED = "cancelled"     // URL param: ?status=cancelled
}

export enum TodoStatus {
  PENDING = "pending",        // URL param: ?status=pending
  IN_PROGRESS = "in_progress", // URL param: ?status=in_progress
  BLOCKED = "blocked",        // URL param: ?status=blocked
  DONE = "done",             // URL param: ?status=done
  CANCELLED = "cancelled"     // URL param: ?status=cancelled
}

export enum TodoType {
  TASK = "task",
  CHECKLIST = "checklist",
  SUBTASK = "subtask"
}

export enum TaskPriority {
  LOW = "low",
  MEDIUM = "medium",
  HIGH = "high",
  URGENT = "urgent"
}

export enum UploadStatus {
  PENDING = "pending",
  PROCESSING = "processing",
  COMPLETED = "completed",
  FAILED = "failed"
}

export enum ChunkUploadStatus {
  UPLOADING = "uploading",
  ASSEMBLING = "assembling",
  COMPLETED = "completed",
  FAILED = "failed",
  ERROR = "error"
}

export enum TodoStatsKey {
  // Status-based counts (from TodoStatus enum)
  TOTAL = "total",                    // URL param: ?filter=total
  PENDING = "pending",                // URL param: ?filter=pending
  IN_PROGRESS = "in_progress",        // URL param: ?filter=in_progress
  BLOCKED = "blocked",               // URL param: ?filter=blocked
  DONE = "done",                      // URL param: ?filter=done
  
  // Time-based computed counts (calculated on the fly - active todos only)
  OVERDUE = "overdue",           // URL param: ?filter=overdue
  DUE_TODAY = "due_today",       // URL param: ?filter=due_today
  COMPLETED_TODAY = "completed_today",  // URL param: ?filter=completed_today
  WITHIN_TIME = "within_time"   // URL param: ?filter=within_time
}

export enum ModuleStatsKey {
  TOTAL = "total",     // Internal stats key for total count
  RECENT = "recent"    // Internal stats key for recent items count
}
